[style] Introduce a RISC-V assembly style guide.

mcy opened the pull request:
This PR is a continuation of #925, which aims to produce an RFC for a RISC-V assembly style guide. I've created this PR to work as a sort of forum driving the RFC forward; see the linked bug for prior discussions.

Fixes #925.

mcy commented:
@imphil @lenary @gkelly @asb for visibility

sjgitty commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+***Avoid inline assembly as much as possible, as long as correctness and readability are not impacted.***
+
+Inline assembly is best reserved for when a high-level language cannot express what we need to do, such as expressing complex control flow or talking to the hardware.
+If a compiler intrinsic can achieve the same effect, such as `__builtin_clz()`; then that should be used instead.
```
comma not semi-colon here?

mcy replied:
Done.

sjgitty commented on doc/rm/asm_coding_style.md:
````diff
@@ -0,0 +1,268 @@
+  ```c
+  asm volatile("wfi");
+  ```
+- Multiple-instruction `asm` statements should be writen with one instruction per line, formatted as follows:
````
"written"

mcy replied:
Done.

sjgitty commented on doc/rm/asm_coding_style.md:
````diff
@@ -0,0 +1,268 @@
+  asm("..." : "=a0"(foo) :: "memory");
+  ```
+
+### Non-returning `asm`.
````
remove trailing `.` as per the other section names?

mcy replied:
Done.

sjgitty commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+***Functions with non-returning `asm` must be marked as `noreturn`.***
+
+Since C and C++ can't tell that an `asm` statement does not yield control, such functions should be endeded with something like
```
"ended with"?

sjgitty replied:
perhaps end the sentence here rather than split with the example in between? like "should be ended with something like the following to avoid undefined behavior".

mcy replied:
Done.

sjgitty commented on doc/rm/asm_coding_style.md:
````diff
@@ -0,0 +1,268 @@
+}
+```
+to avoid undefined behavior.
+Note that infinite loops without side effects are UB, so your infinite loop should contain `volatile` inline assembly to trick the compiler otherwise.
````
UB? Am I missing a known SWism?

sjgitty replied:
Minor: this is only the second time second-person is used. Perhaps "so infinite loops should contain" rather than "your infinite loop"

imphil replied:
"undefined behavio(u)r", but please spell it out

mcy replied:
Hehe, yeah "UB" is a very common term of art in language-lawyering, so I tend to forget to expand it out.

lenary commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+### Register Names
+
+***When referring to a RISC-V register, they must be referred to by their ABI names.***
```
Yes, here: https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#register-convention (this is the psABI specification document).

mcy replied:
Added.

mcy commented:
Everyone's comments should be addressed now.

imphil commented on doc/rm/asm_coding_style.md:
````diff
@@ -0,0 +1,275 @@
+  ```c
+  asm volatile(
+    "  la sp, _stack_start;"
+    "  tail _crt0;         "
````
What's the purpose of having that many spaces before the closing `"`? Seems to make editing harder for no real readability gain IMHO.

mcy replied:
I'll get rid of it if other people don't like it. Specifically, I'm just imitating the standard multi-line macro style:
```c
#define MY_MACRO \
  int x = 0;                   \
  return x;
```

lenary replied:
Yeah, let's remove trailing whitespace from assembly string literals. Leading whitespace is useful and should be kept.

mcy replied:
Fine, fine. I've also added an example label, and formatted the instruction operands to start on the same column.

lenary commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,275 @@
+
+This style guide is specialized for R32IMC, the ISA implemented by Ibex. 
+As such, no advice is provided for other RISC-V extensions, though this style guide is written such that advice for other extensions could be added without conflicts.
+
```
The style guide should also include the "Terminology Conventions" sections as present in the other style guides.

mcy replied:
Funny enough, it originally did, and I was asked to remove it. I think it was either @imphil or @asb? I might be misremembering.

lenary replied:
Ok, no worries.

mcy commented:
@lenary your comments should be addressed now.

luismarques commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,279 @@
+***Jumping into a C function must be done either with a `call` instruction, or, if that function is marked `noreturn`, a `tail` instruction.***
+
+Doing an unconditional jump with `j` into a C function can lead to erratic behavior if the C compiler emits a `ret` instruction for it.
+Furthermore, calling across translation units should be performed with a far jump, like `call` or `tail`, to avoid linking woes.
```
Can we please be a bit more explicit here about what the problem is when you use the `j` and how `call` and `tail` solves it? Perhaps giving an example of how when you change how things are linked the `j` approach can break. Also, the meaning of a "far jump" might not be obvious to the reader, so consider something more explicit (e.g. "a pseudo-instruction that expands to the correct sequence", etc.)

mcy replied:
I've revised that paragraph. Let me know what you think.

luismarques commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,279 @@
+
+***Functions with non-returning `asm` must be marked as `noreturn`.***
+
+C and C++ compilers are, in general, not supposed to introspect `asm` blocks, and as such cannot deterimine that they never return.
```
Typo ("deterimine").

mcy replied:
Done.

lenary commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+***Comments must use either the `//` or `/* */` syntaces.***
+
+Every function-like label which is meant to be called like a function (*especially* `.globl`s) should be given a Doxygen-style comment.
+While Doxygen is not suited for assembly, that style should be used for consistency.
```
Done,

lenary commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,275 @@
+Every function-like label which is meant to be called like a function (*especially* `.globl`s) should be given a Doxygen-style comment.
+While Doxygen is not suited for assembly, that style should be used for consistency.
+
+All other advice for writing comments, as in the C/C++ style guide, also applies.
```
Done.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+### Summary
+
+OpenTitan needs to implement substancial functionality directly in RISC-V assembly.
```
Fixed.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+### Pseudoinstrucions
+
+***When performing an operating for which a pseudoinstruction exists, that pseudoinstruction must be used.***
```
Done.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+***When performing an operating for which a pseudoinstruction exists, that pseudoinstruction must be used.***
+
+Pseudoinstructions make RISC-V's otherwise verbose RISC style more readable; for
```
Done.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+  sb t3, 0(a0)
+  // Wrong:
+  lw t3, sp, 8
+  sb t3, sp
```
Done.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+### Comments
+
+***Comments must use either the `//` or `/* */` syntaces.***
```
Done.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,268 @@
+
+### Register useage
+
+***Register usage in a "function" that diverges from the RISC-V function call ABI *must* be documented.***
```
Done.

asb commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,286 @@
+
+While Ibex implements the RISC-V C extension, it is expected that the toolchain will automatically compress instructions where possible.
+This advice extends to the `c.unimp` all-zeroes illegal instruction.
+Toolchains like `LLVM` and `gas` are expected to recognize the `unimp` mnemonic to stand for two `c.unimp` instructions.
```
Actually, `unimp` is compressed to a single `c.unimp`. These two lines around `unimp` perhaps aren't necessary in this doc?

mcy replied:
Done.

asb commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,286 @@
+
+***Register usage in a "function" that diverges from the RISC-V function call ABI must be documented.***
+
+This includes non-standard calling conventions, non-standard clobbers, and other behavior not expected of a well-behaved RISV-V function.
```
Don't we want similar annotations for assembly functions that do match the RISC-V function call ABI too?

I think it would also be worth adding some guidance on annotating register usage within a function too. Maybe something like "Although lengthy assembly functions are undesirable, if they are necessary care should be taken to annotate the assignment of logical values to registers within the function in a way that makes it easier to understand."?

mcy replied:
Done.

asb commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,286 @@
+***Every code path within an assembly file must end in a non-linking jump.***
+
+Assembly should be written such that the program counter can't wander off past the written instructions.
+As such, all assembly should be ended with `ret` (or any of the protection ring returns like `mret`), an infinite `wfi` loop, or an instruction that is guaranteed to trap and not return, like an `exit`-like syscall.
```
If we're happy with `unimp` here, probably worth listing explicitly?

mcy replied:
Done.

asb commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,286 @@
+
+***Register usage in a "function" that diverges from the RISC-V function call ABI must be documented.***
+
+This includes non-standard calling conventions, non-standard clobbers, and other behavior not expected of a well-behaved RISV-V function.
```
RISV-V => RISC-V

mcy replied:
Oops.

mcy commented:
@asb I've addressed your comments.

Furthermore, I've added some additional formatting guidance, which I feel might be a bit convoluted. Basically, it requires that instruction operands be tab (two-space) aligned and on the same column, but I think it might not be wrong to declare that all our inline assembly be spaced like this:
```S
my_label:
  add     t0, t1, t2
  call    my_function
  really.long.op ra
  [mnemo] [operands...]
```
(I.e., two spaces, followed by seven columns for the mnemonic, a space, and then the operands. Thus, instructions start at column 2, and operands at column 10.) Thoughts?

asb commented:
The new changes look good to me, thanks.

Regarding aligning the operands to an offset based on the longest mnemonic: my personal taste is to not bother and just have `mnemonic op1, op2, op3`. This looks fine to my eyes, and doesn't have the problem of worrying about re-aligning operands when adding new lines with longer mnemonics etc.

mcy commented:
Yeah, that's what I was thinking- hence why I propose instead having a fixed "table" style, instead.

asb commented:
The table solution and the requested alignment in the current text don't seem worth it IMHO. My personal preference is to not worry about trying to align the operands, and just have a single space from mnemonic to operand.

mcy commented:
Sure, I guess. I'll remove it and not bother.

I might write a simple assembly formatter... but that's a PInfinity right now.

asb commented:
You've got a typo in your commit title, Introdice => Introduce. You should also add a link to this guide to `doc/rm/_index.md`

mcy commented:
Fixed.

mcy commented:
Alex suggests in #925 that we should get this landed, since the RFC process won't get instantiated for some time and this change seems to already have roughly unanimous positive reactions from stakeholders; I'll let one more round of reviews happen before moving on to merging.

asb commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,279 @@
+title: "RISC-V Assembly Style Guide"
+---
+
+## Basics
```
I'm not sure we want to do that - the title will be rendered as a `<h1>`, so it makes more sense for these sub-headings to be `##` and so render as `<h2>`?

gkelly replied:
Hmm, maybe we should be generating a different element for the title then.  It feels strange to change the markdown to accommodate styling.

mcy commented on doc/rm/asm_coding_style.md:
```diff
@@ -0,0 +1,305 @@
+
+See the [psABI
+Reference](https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#register-convention)
+for a reference to these names.
```
Done.
